### DISCUSSION TITLE
Implementing HID emulation

### DISCUSSION THREAD

**Discussion by yuvadm** [bluez/bluer#102] [Q&A] [answered]
What's the easiest API to use to emulate an HID device? e.g. I'd like my library running on a Linux box to emulate a Bluetooth keyboard. Should I be using GATT?

Maximum compatibility would be nice, but modern smartphone devices can be assumed to be my target devices, so ease of implementation is preferred.

**Comment by surban** [comment]
If your target device, i.e. the device connecting to your emulated keyboard, supports HID over Bluetooth Low Energy, you should implement https://www.bluetooth.com/specifications/specs/human-interface-device-service-1-0/ .

**Reply by yuvadm** [reply]
Thanks @surban , seems like most modern implementations indeed do HID over BLE, but I can't find many examples that use bluez APIs. Any chance you can point me to some starting points? I'd eventually love to contribute back some basic examples to be included in this repo.

**Reply by surban** [reply, marked as answer]
I would recommend that you start with the [GATT server example](https://github.com/bluez/bluer/blob/master/bluer/examples/gatt_server_io.rs) and modify it to implement the minimum required services and characteristics defined by the [Bluetooth HID specification](https://www.bluetooth.com/specifications/specs/human-interface-device-service-1-0/). For that you will also need to take a look at the [USB HID specification](https://www.usb.org/sites/default/files/hid1_11.pdf), as it defines the actual HID events and commands.

**Reply by yuvadm** [reply]
Awesome, thanks for all the references!
